libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridFromSplines.cpp

```diff
@@ -592,11 +643,335 @@ namespace meshkernel
+        }
+    }
+
+    double CurvilinearGridFromSplines::ComputeNodeSegmentCrossingTime(const Point& x1, const Point& x3, const Point& x4,
+                                                                      const Point& v1, const Point& v3, const Point& v4) const
+    {
+        if ((x1.y - x3.y) * (x4.x - x3.x) - (x1.x - x3.x) * (x4.y - x3.y) < 0.0)
+        {
+            return 1.0e99;
+        }
+
+        // Compute differences
+        Point x13 = x3 - x1;
+        Point x34 = x4 - x3;
+        Point v13 = v3 - v1;
+        Point v34 = v4 - v3;
+
+        double a = cross(v13, v34);
+        double b = cross(x13, v34) - cross(x34, v13);
+        double c = cross(x13, x34);
+
+        std::array<double, MaxDegreeP1> coeffs{a, b, c};
+        std::array<double, MaxDegree> roots{0.0, 0.0};
+        std::array<double, MaxDegree> beta;
+        beta.fill(constants::missing::doubleValue);
+        roots.fill(constants::missing::doubleValue);
+
+        SolveQuadratic(coeffs, roots);
+
+        for (UInt i = 0; i < MaxDegree; ++i)
+        {
+
+            if (roots[i] == constants::missing::doubleValue)
+            {
+                continue;
+            }
+
+            if (roots[i] < tolerance)
+            {
+                continue;
+            }
+            Point xs = x4 - x3 + (v4 - v3) * roots[i];
+            double det = length(xs);
+
+            if (std::abs(det) > tolerance)
+            {
+                beta[i] = -dot(x3 - x1 + (v3 - v1) * roots[i], xs) / det;
+            }
+        }
+
+        double time = 1.0e99;
+
+        for (UInt i = 0; i < MaxDegree; ++i)
+        {
+            // Extend limits by a relatively small amount
+            const double lowerLimit = -8.0 * std::numeric_limits<double>::epsilon();
+            const double upperLimit = 1.0 + 8.0 * std::numeric_limits<double>::epsilon();
+
+            if (beta[i] >= lowerLimit && beta[i] <= upperLimit && roots[i] >= 0.0 && roots[i] != constants::missing::doubleValue)
+            {
+                time = std::min(time, roots[i]);
+            }
+        }
+
+        if (time == constants::missing::doubleValue || time <= 0.0)
+        {
+            time = 1.0e99;
+        }
+
+        return time;
+    }
+
+    bool CurvilinearGridFromSplines::IncludeDirectNeighbours(const UInt layerIndex,
+                                                             const UInt loopIndex,
+                                                             const lin_alg::Matrix<UInt>& gridPointsIndices,
+                                                             const UInt indexLeftOfLeft,
+                                                             const UInt indexRightOfRight) const
+    {
+
+        UInt i1 = gridPointsIndices(loopIndex, 0);
+        UInt i2 = gridPointsIndices(loopIndex + 1, 0);
+        UInt j1 = gridPointsIndices(loopIndex, 1);
+        UInt j2 = gridPointsIndices(loopIndex + 1, 1);
+
+        if (indexRightOfRight >= indexLeftOfLeft)
```

Comment:
// Extract grid point indices
    const UInt i1 = gridPointsIndices(loopIndex, 0);
    const UInt i2 = gridPointsIndices(loopIndex + 1, 0);
    const UInt j1 = gridPointsIndices(loopIndex, 1);
    const UInt j2 = gridPointsIndices(loopIndex + 1, 1);

    // Common check for layer index validity
    if (j1 < layerIndex - 1 || j2 < layerIndex - 1)
    {
        return true;  // If layer indices are out of bounds, the neighbors should be included
    }

    // Check index range conditions based on the relationship between indexRightOfRight and indexLeftOfLeft
    const bool isWithinRange = (indexRightOfRight >= indexLeftOfLeft)
                               ? ((i1 > indexLeftOfLeft && i1 < indexRightOfRight) || (i2 > indexLeftOfLeft && i2 < indexRightOfRight))
                               : (!((i1 >= indexRightOfRight && i1 <= indexLeftOfLeft) || (i2 >= indexRightOfRight && i2 <= indexLeftOfLeft)));

    // If the points are within the range, exclude the direct neighbors
    return !isWithinRange;

Reply:
this seems to work